src/main/java/io/cdap/plugin/ariba/source/AribaServices.java

```diff
@@ -719,14 +741,17 @@ public boolean isApiLimitExhausted(Response response) {
    * Check for limit and status code than throws exception accordingly
    *
    * @param response response
+   * @param shouldWait Do we to wait for time defined in header if API limit is exhausted
    * @throws AribaException
    * @throws InterruptedException
    */
   @VisibleForTesting
-  void checkAndThrowException(Response response) throws AribaException, InterruptedException {
+  void checkAndThrowException(Response response, boolean shouldWait) throws AribaException, InterruptedException,
+    AribaRetryableException {
     if (response.code() == HttpURLConnection.HTTP_BAD_REQUEST && !AribaUtil.isNullOrEmpty(response.message())) {
       throw new AribaException(response.message(), response.code());
     }
+
     boolean limitExhausted = isApiLimitExhausted(response);
 
     if (limitExhausted && isDayLimitExhausted) {
```

Comment:
where are we resetting these variables `isDayLimitExhausted`, `isHourLimitExhausted`, `isMinuteLimitExhausted` and `isSecondsLimitExhausted`??

Consider a use case,

where on first retry day limit was exhausted but on the second retry only hour limit was exhausted but because day limit exhausted was set to true earlier, it will still execute this code path?

Wdyt?

Generally, we should use limited scope variables for such cases.

Reply:
This case is not rely possible, as for `isDayLimitExhausted` case we do not wait but throws an error, this was suggested by albert as this will consume resources on dataproc for no reason.

Anyway i can see why you suggested this, if you want i can add this line `isApiLimitExhausted` function.
```java 
isDayLimitExhausted = isHourLimitExhausted = isMinuteLimitExhausted = isSecondsLimitExhausted = false;
```

Reply:
Thanks, can we please also add a unit test for this case.

For eg, first response results `HourLimitExhausted` and second response returns `MinuteLimitExhausted`.

Reply:
Adding this !

For second iteration i check the the flags were reset properly using
```java
Assert.assertTrue(aribaServices.isMinuteLimitExhausted);
Assert.assertFalse(aribaServices.isHourLimitExhausted);
```

```java
 @Test
  public void testApiLimitExhaustedFlagsReset() {
    Response hourLimitExhaustedResponse = new Response.Builder()
      .request(new Request.Builder().url("https://some-url.com").build())
      .protocol(Protocol.HTTP_2)
      .code(429)
      .message("Hourly limit exceeded")
      .header("X-RateLimit-Remaining-Day", "1")
      .header("X-RateLimit-Remaining-Hour", "0")
      .body(ResponseBody.create(MediaType.parse("application/json"), "{}"))
      .build();

    Assert.assertTrue(aribaServices.isApiLimitExhausted(hourLimitExhaustedResponse));
    Assert.assertTrue(aribaServices.isHourLimitExhausted);

    Response minuteLimitExhaustedResponse = new Response.Builder()
      .request(new Request.Builder().url("https://some-url.com").build())
      .protocol(Protocol.HTTP_2)
      .code(429)
      .message("Minute limit exceeded")
      .header("X-RateLimit-Remaining-Day", "1")
      .header("X-RateLimit-Remaining-Hour", "1")
      .header("X-RateLimit-Remaining-Minute", "0")
      .body(ResponseBody.create(MediaType.parse("application/json"), "{}"))
      .build();

    Assert.assertTrue(aribaServices.isApiLimitExhausted(minuteLimitExhaustedResponse));
    Assert.assertTrue(aribaServices.isMinuteLimitExhausted);
    Assert.assertFalse(aribaServices.isHourLimitExhausted);
  }
  ```